packages/logger/src/lib/Logger.ts

```diff
@@ -0,0 +1,194 @@
+import { Logger as BuiltinLogger, LogLevel, LogMethods } from '@sapphire/framework';
+import { bgRed, cyan, gray, magenta, options as coloretteOptions, red, Style, white, yellow } from 'colorette';
+import { Console } from 'console';
+import { inspect, InspectOptions } from 'util';
+import { LoggerLevel, LoggerLevelOptions } from './LoggerLevel';
+
+/**
+ * The logger class.
+ * @since 1.0.0
+ */
+export class Logger extends BuiltinLogger {
+	/**
+	 * The console this writes to.
+	 * @since 1.0.0
+	 */
+	protected readonly console: Console;
+
+	/**
+	 * The formats supported by the logger.
+	 * @since 1.0.0
+	 */
+	protected readonly formats: Map<LogLevel, LoggerLevel>;
+
+	/**
+	 * The string write will join values by.
+	 * @since 1.0.0
+	 */
+	protected readonly join: string;
+
+	public constructor(options: LoggerOptions = {}) {
+		super(options.level ?? LogLevel.Info);
+
+		this.console = new Console(options.stdout ?? process.stdout, options.stderr ?? process.stderr);
+		this.formats = Logger.createFormatMap(options.format, options.defaultFormat);
+		this.join = options.join ?? ' ';
+
+		if (typeof options.colors === 'boolean') coloretteOptions.enabled = options.colors;
+	}
+
+	/**
+	 * Writes the log message given a level and the value(s).
+	 * @param level The log level.
+	 * @param values The values to log.
+	 */
+	public write(level: LogLevel, ...values: readonly unknown[]): void {
+		if (level < this.level) return;
+
+		const method = this.levels.get(level) ?? 'log';
+		const formatter = this.formats.get(level) ?? this.formats.get(LogLevel.None)!;
+
+		this.console[method](formatter.run(this.preprocess(values)));
+	}
+
+	/**
+	 * Pre-processes an array of values.
+	 * @since 1.0.0
+	 * @param values The values to pre-process.
+	 */
+	protected preprocess(...values: readonly unknown[]) {
+		const inspectOptions: InspectOptions = { colors: coloretteOptions.enabled, depth: 0 };
+		return values.map((value) => (typeof value === 'string' ? value : inspect(value, inspectOptions))).join(this.join);
+	}
+
+	private get levels() {
+		return Reflect.get(BuiltinLogger, 'levels') as Map<LogLevel, LogMethods>;
+	}
+
+	private static createFormatMap(options: LoggerFormatOptions = {}, defaults: LoggerLevelOptions = options.none ?? {}) {
+		const map = new Map<LogLevel, LoggerLevel>();
+
+		map.set(LogLevel.Trace, Logger.ensureDefaultLevel(options.trace, defaults, gray, 'TRACE'));
+		map.set(LogLevel.Debug, Logger.ensureDefaultLevel(options.debug, defaults, magenta, 'DEBUG'));
+		map.set(LogLevel.Info, Logger.ensureDefaultLevel(options.info, defaults, cyan, 'INFO'));
+		map.set(LogLevel.Warn, Logger.ensureDefaultLevel(options.warn, defaults, yellow, 'WARN'));
+		map.set(LogLevel.Error, Logger.ensureDefaultLevel(options.error, defaults, red, 'ERROR'));
+		map.set(LogLevel.Fatal, Logger.ensureDefaultLevel(options.fatal, defaults, bgRed, 'FATAL'));
+		map.set(LogLevel.None, Logger.ensureDefaultLevel(options.none, defaults, white, ''));
+
+		return map;
```

Comment:
```suggestion
		return new Map<LogLevel, LoggerLevel>([
			[LogLevel.Trace, Logger.ensureDefaultLevel(options.trace, defaults, gray, 'TRACE')],
			[LogLevel.Debug, Logger.ensureDefaultLevel(options.debug, defaults, magenta, 'DEBUG')],
			[LogLevel.Info, Logger.ensureDefaultLevel(options.info, defaults, cyan, 'INFO')],
			[LogLevel.Warn, Logger.ensureDefaultLevel(options.warn, defaults, yellow, 'WARN')],
			[LogLevel.Error, Logger.ensureDefaultLevel(options.error, defaults, red, 'ERROR')],
			[LogLevel.Fatal, Logger.ensureDefaultLevel(options.fatal, defaults, bgRed, 'FATAL')],
			[LogLevel.None, Logger.ensureDefaultLevel(options.none, defaults, white, '')]
		]);
```

Reply:
Or in regular TS because suggestion formats it horribly:
```ts
	return new Map<LogLevel, LoggerLevel>([
		[LogLevel.Trace, Logger.ensureDefaultLevel(options.trace, defaults, gray, 'TRACE')],
		[LogLevel.Debug, Logger.ensureDefaultLevel(options.debug, defaults, magenta, 'DEBUG')],
		[LogLevel.Info, Logger.ensureDefaultLevel(options.info, defaults, cyan, 'INFO')],
		[LogLevel.Warn, Logger.ensureDefaultLevel(options.warn, defaults, yellow, 'WARN')],
		[LogLevel.Error, Logger.ensureDefaultLevel(options.error, defaults, red, 'ERROR')],
		[LogLevel.Fatal, Logger.ensureDefaultLevel(options.fatal, defaults, bgRed, 'FATAL')],
		[LogLevel.None, Logger.ensureDefaultLevel(options.none, defaults, white, '')]
	]);
```